File: lab/python/ordenacoes/merge_sort.py
def merge(esquerda, direita):
    resultado = []
    i = j = 0

    while i < len(esquerda) and j < len(direita):
        if esquerda[i] < direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1

    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado


def merge_sort_iterativo(lista):
    n = len(lista)
    tamanho = 1  # começa juntando blocos de 1 elemento

    while tamanho < n:
        for i in range(0, n, 2 * tamanho):
            meio = i + tamanho
            fim = i + 2 * tamanho
            esquerda = lista[i:meio]
            direita = lista[meio:fim]
            lista[i:fim] = merge(esquerda, direita)
        tamanho *= 2  # dobra o tamanho do bloco a cada rodada

    return lista


valores = [8, 3, 5, 2, 9, 1]
print("Antes:", valores)
print("Depois:", merge_sort_iterativo(valores))

File: lab/javascript/ordenacoes/merge_sort.js
function merge(esquerda, direita) {
  const resultado = [];
  let i = 0, j = 0;

  while (i < esquerda.length && j < direita.length) {
    if (esquerda[i] < direita[j]) {
      resultado.push(esquerda[i++]);
    } else {
      resultado.push(direita[j++]);
    }
  }
  // cola o resto
  return resultado.concat(esquerda.slice(i)).concat(direita.slice(j));
}

function merge_sort_iterativo(lista) {
  const n = lista.length;
  let tamanho = 1; // blocos de 1

  while (tamanho < n) {
    for (let i = 0; i < n; i += 2 * tamanho) {
      const meio = i + tamanho;
      const fim = Math.min(i + 2 * tamanho, n);
      const esquerda = lista.slice(i, Math.min(meio, n));
      const direita  = lista.slice(Math.min(meio, n), fim);
      const merged = merge(esquerda, direita);
      // substitui o trecho [i, fim) pelo merge
      lista.splice(i, fim - i, ...merged);
    }
    tamanho *= 2;
  }
  return lista;
}

// ---- teste
let valores = [8, 3, 5, 2, 9, 1];
console.log("Antes:", valores.slice());
console.log("Depois:", merge_sort_iterativo(valores));
